alternator/ttl.cc

```diff
@@ -113,4 +139,321 @@ future<executor::request_return_type> executor::describe_time_to_live(client_sta
+static future<> expire_item(service::storage_proxy& proxy,
+                            const service::query_state& qs,
+                            const std::vector<bytes_opt>& row,
+                            const std::vector<unsigned>& pk_columns,
+                            const std::vector<unsigned>& ck_columns,
+                            schema_ptr schema,
+                            api::timestamp_type ts) {
+    // Prepare the row key to delete
+    std::vector<bytes> exploded_pk;
+    for (unsigned c : pk_columns) {
+        const auto& row_c = row[c];
+        if (!row_c) {
+            // This shouldn't happen - all key columns must have values.
+            // But if it ever happens, let's just *not* expire the item.
+            return make_ready_future<>();
+        }
+        exploded_pk.push_back(*row_c);
+    }
+    auto pk = partition_key::from_exploded(exploded_pk);
+    mutation m(schema, pk);
+    // If there's no clustering key, a tombstone should be created directly
+    // on a partition, not on a clustering row - otherwise it will look like
+    // an open-ended range tombstone, which will crash on KA/LA sstable format.
+    // See issue #6035
+    if (ck_columns.empty()) {
+        m.partition().apply(tombstone(ts, gc_clock::now()));
+    } else {
+        std::vector<bytes> exploded_ck;
+        for (unsigned c : ck_columns) {
+            const auto& row_c = row[c];
+            if (!row_c) {
+                // This shouldn't happen - all key columns must have values.
+                // But if it ever happens, let's just *not* expire the item.
+                return make_ready_future<>();
+            }
+            exploded_ck.push_back(*row_c);
+        }
+        auto ck = clustering_key::from_exploded(exploded_ck);
+        m.partition().clustered_row(*schema, ck).apply(tombstone(ts, gc_clock::now()));
+    }
+    return proxy.mutate(std::vector<mutation>{std::move(m)},
+        db::consistency_level::LOCAL_QUORUM,
+        executor::default_timeout(), // FIXME - which timeout?
+        qs.get_trace_state(), qs.get_permit());
+}
+
+
+future<> expiration_service::run() {
+    // FIXME: don't just tight-loop, think about timing, pace, and
+    // store position in durable storage, etc.
+    // FIXME: If one of the iterations below fails with an exception,
+    // this service will die forever. It's probably not a good idea!
+    // We should log the failure, but not stop the service.
+    for (;;) {
+        // FIXME: think about working on different tables in parallel.
+        // also need to notice when a new table is added, a table is
+        // deleted or when ttl is enabled or disabled for a table!
+        for (const auto& cf : _db.get_column_families()) {
+            schema_ptr s = cf.second->schema();
+            // Check if an expiration-time attribute is enabled for this
+            // table.
+            std::optional<std::string> attribute_name = find_tag(*s, TTL_TAG_KEY);
```

Comment:
From caller's perspective, it looks like `find_tag` could just as well return an `optional<string_view>` and not own the string, which would save us a copy, since we never hold this `attribute_name` across a preemption point, and thus the string view can't get invalidated. On the other hand, these strings are most likely very small and perhaps it's better to obtain a copy here and make this function safe even if we want to use `attribute_name` across a preemption point in the future.

Reply:
If find_tag returns a string, it is just moved in this line 279, and not copied again. You're right that if find_tag returned a string_view and this code also saved a string_view, one copy would be saved. But it looks very scary - the pointer we're saving isn't to some object which the caller is holding (e.g., some document we're working on) - but rather to some transient list of tags. If a preemption point does happen - this pointer can get invalidated. So it seems like it's a bad idea to risk it - especially because this is not a frequent operation (this is just for checking once in a while on which attribute the ttl is enabled).